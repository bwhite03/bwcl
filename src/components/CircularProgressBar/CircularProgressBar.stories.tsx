import React from 'react';
import { Story, Meta } from '@storybook/react';
import CircularProgressBar, {
  CircularProgressProps,
} from './CircularProgressBar';

export default {
  title: 'CircularProgressBar',
  component: CircularProgressBar,
} as Meta;

const Template: Story<CircularProgressProps> = (args) => (
  <CircularProgressBar {...args} />
);

export const CircularProgressBarExample = Template.bind({});
CircularProgressBarExample.args = {
  size: 150,
  progress: 75,
  strokeWidth: 15,
  circleStrokeOne: '#d93dfe',
  showPercent: false,
};
