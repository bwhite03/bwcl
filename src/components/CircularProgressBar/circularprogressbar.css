.circular-progress-bar {
  display: block;
  margin: 20px auto;
  max-width: 100%;
  max-height: 250px;
  margin-top: 100px;
  background: transparent;
  color: #fff;
}

.circular-progress-bar-bg {
  fill: none;
}

.circular-progress-bar-circle {
  fill: none;
}

.circular-progress-bar-percentage {
  font-size: 3rem;
  text-anchor: middle;
  fill: #fff;
  font-weight: bold;
  transform: translateY(10px);
}
