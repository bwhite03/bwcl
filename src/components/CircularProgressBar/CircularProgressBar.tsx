import React, { useState, useEffect, useRef } from 'react';
import './circularprogressbar.css';

export interface CircularProgressProps {
  size: number;
  progress: number;
  strokeWidth: number;
  circleStrokeOne: string;
  circleStrokeTwo?: string;
  showPercent: boolean;
}

const CircularProgress = (props: CircularProgressProps) => {
  const [offset, setOffset] = useState(0);
  const circleRef = useRef<SVGCircleElement>(null);
  const [progressText, setProgressText] = useState('');
  const [progressColor, setProgressColor] = useState('');

  const { size, progress, strokeWidth, circleStrokeOne, showPercent } = props;

  const center = size / 2;
  const radius = size / 2 - strokeWidth / 2;
  const circumference = 2 * Math.PI * radius;

  useEffect(() => {
    const progressOffSet = ((100 - progress) / 100) * circumference;
    setProgressText(progress.toString());
    setOffset(progressOffSet);

    if (circleRef.current) {
      // @ts-ignore
      circleRef.current.style =
        'transition: stroke-dashoffset 850ms ease-in-out, stroke 500ms linear';
    }

    if (progress >= 75) {
      setProgressColor('#47e61c');
    } else if (progress < 25) {
      setProgressColor('#f00e0e');
    } else if (progress < 50) {
      setProgressColor('#f59b33');
    } else if (progress < 75) {
      setProgressColor('#f00e0e');
    }
  }, [setOffset, progress, circumference, offset]);

  return (
    <div>
      <svg className="circular-progress-bar" width={size} height={size}>
        <circle
          className="circular-progress-bar-bg"
          stroke={circleStrokeOne}
          cx={center}
          cy={center}
          r={radius}
          strokeWidth={strokeWidth}
          fill="none"
        />
        <circle
          ref={circleRef}
          className="circular-progress-bar-circle"
          stroke={progressColor}
          cx={center}
          cy={center}
          r={radius}
          strokeWidth={strokeWidth}
          strokeDasharray={circumference}
          strokeDashoffset={offset}
          fill="none"
        />
        <text
          x={center}
          y={center}
          className="circular-progress-bar-percentage"
        >
          {parseInt(progressText.toString() || '30')}
          {showPercent ? '%' : null}
        </text>
      </svg>
    </div>
  );
};

export default CircularProgress;
