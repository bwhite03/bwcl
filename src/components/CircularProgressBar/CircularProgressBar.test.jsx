import React from 'react';
import { render } from '@testing-library/react';
import CircularProgressBar from './CircularProgressBar';

describe('CircularProgressBar', () => {
  test('should render correctly', () => {
    const { container } = render(
      <CircularProgressBar
        progress={25}
        size={150}
        strokeWidth={15}
        circleStrokeOne={'#ff0000'}
        showPercent={false}
      />
    );
    expect(container).toMatchSnapshot();
  });

  test('should show percent', () => {
    const { container } = render(
      <CircularProgressBar
        progress={25}
        size={150}
        strokeWidth={15}
        circleStrokeOne={'#ff0000'}
        showPercent={true}
      />
    );
    expect(container).toMatchSnapshot();
  });

  test('should render with score > 75', () => {
    const { container } = render(
      <CircularProgressBar
        progress={90}
        size={150}
        strokeWidth={15}
        circleStrokeOne={'#ff0000'}
        showPercent={true}
      />
    );
    expect(container).toMatchSnapshot();
  });

  test('should render with score < 25', () => {
    const { container } = render(
      <CircularProgressBar
        progress={15}
        size={150}
        strokeWidth={15}
        circleStrokeOne={'#ff0000'}
        showPercent={true}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
