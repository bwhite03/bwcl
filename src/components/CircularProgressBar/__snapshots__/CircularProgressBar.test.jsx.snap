// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CircularProgressBar should render correctly 1`] = `
<div>
  <div>
    <svg
      class="circular-progress-bar"
      height="150"
      width="150"
    >
      <circle
        class="circular-progress-bar-bg"
        cx="75"
        cy="75"
        fill="none"
        r="67.5"
        stroke="#ff0000"
        stroke-width="15"
      />
      <circle
        class="circular-progress-bar-circle"
        cx="75"
        cy="75"
        fill="none"
        r="67.5"
        stroke="#f59b33"
        stroke-dasharray="424.11500823462205"
        stroke-dashoffset="318.0862561759665"
        stroke-width="15"
        style="transition: stroke-dashoffset 850ms ease-in-out, stroke 500ms linear;"
      />
      <text
        class="circular-progress-bar-percentage"
        x="75"
        y="75"
      >
        25
      </text>
    </svg>
  </div>
</div>
`;

exports[`CircularProgressBar should render with score < 25 1`] = `
<div>
  <div>
    <svg
      class="circular-progress-bar"
      height="150"
      width="150"
    >
      <circle
        class="circular-progress-bar-bg"
        cx="75"
        cy="75"
        fill="none"
        r="67.5"
        stroke="#ff0000"
        stroke-width="15"
      />
      <circle
        class="circular-progress-bar-circle"
        cx="75"
        cy="75"
        fill="none"
        r="67.5"
        stroke="#f00e0e"
        stroke-dasharray="424.11500823462205"
        stroke-dashoffset="360.49775699942876"
        stroke-width="15"
        style="transition: stroke-dashoffset 850ms ease-in-out, stroke 500ms linear;"
      />
      <text
        class="circular-progress-bar-percentage"
        x="75"
        y="75"
      >
        15
        %
      </text>
    </svg>
  </div>
</div>
`;

exports[`CircularProgressBar should render with score < 75 1`] = `
<div>
  <div>
    <svg
      class="circular-progress-bar"
      height="150"
      width="150"
    >
      <circle
        class="circular-progress-bar-bg"
        cx="75"
        cy="75"
        fill="none"
        r="67.5"
        stroke="#ff0000"
        stroke-width="15"
      />
      <circle
        class="circular-progress-bar-circle"
        cx="75"
        cy="75"
        fill="none"
        r="67.5"
        stroke="#daf542"
        stroke-dasharray="424.11500823462205"
        stroke-dashoffset="127.23450247038662"
        stroke-width="15"
        style="transition: stroke-dashoffset 850ms ease-in-out, stroke 500ms linear;"
      />
      <text
        class="circular-progress-bar-percentage"
        x="75"
        y="75"
      >
        70
        %
      </text>
    </svg>
  </div>
</div>
`;

exports[`CircularProgressBar should render with score > 75 1`] = `
<div>
  <div>
    <svg
      class="circular-progress-bar"
      height="150"
      width="150"
    >
      <circle
        class="circular-progress-bar-bg"
        cx="75"
        cy="75"
        fill="none"
        r="67.5"
        stroke="#ff0000"
        stroke-width="15"
      />
      <circle
        class="circular-progress-bar-circle"
        cx="75"
        cy="75"
        fill="none"
        r="67.5"
        stroke="#47e61c"
        stroke-dasharray="424.11500823462205"
        stroke-dashoffset="42.411500823462205"
        stroke-width="15"
        style="transition: stroke-dashoffset 850ms ease-in-out, stroke 500ms linear;"
      />
      <text
        class="circular-progress-bar-percentage"
        x="75"
        y="75"
      >
        90
        %
      </text>
    </svg>
  </div>
</div>
`;

exports[`CircularProgressBar should show percent 1`] = `
<div>
  <div>
    <svg
      class="circular-progress-bar"
      height="150"
      width="150"
    >
      <circle
        class="circular-progress-bar-bg"
        cx="75"
        cy="75"
        fill="none"
        r="67.5"
        stroke="#ff0000"
        stroke-width="15"
      />
      <circle
        class="circular-progress-bar-circle"
        cx="75"
        cy="75"
        fill="none"
        r="67.5"
        stroke="#f59b33"
        stroke-dasharray="424.11500823462205"
        stroke-dashoffset="318.0862561759665"
        stroke-width="15"
        style="transition: stroke-dashoffset 850ms ease-in-out, stroke 500ms linear;"
      />
      <text
        class="circular-progress-bar-percentage"
        x="75"
        y="75"
      >
        25
        %
      </text>
    </svg>
  </div>
</div>
`;
